torch/_dynamo/guards.py

```diff
@@ -542,6 +542,7 @@ def source_ref(source):
                 # TODO: we could make use of 'DefaultsSource' and offer a .guard.is_defaults() API
                 and "__defaults__" not in guard.name
                 and "__kwdefaults__" not in guard.name
+                and "hooks" not in guard.name
```

Comment:
@voznesenskym line 541 checks some property (`and guard.is_nn_module()`) which i assumed would solve the issue you raised in workchat: "hooks" not in [guard.name](http://guard.name/) is too loose, I could pass you a tensor named "my_hooks" and it would go unguarded"

but if you are right, then i think this is the easy part to fix (like you suggested, perhaps with a new source or something)

my main concern is the cost of these guards.

Reply:
Ah okay, this is for the exclusion logic - I misunderstood where this was. (I think I saw a partial snippet)

So this is probably okay - but I am surprised that it adds that many hooks - is it 1 per module? where is the 8-10 number coming from?

Reply:
I didn't test nesting of modules here.  So this is just for 1 module with 1 hook.

The formula should be something like 
  N guards per module (for checking the keys match all the hook  dicts
  M additional guards per module, where that module has M hooks registered 

```
[2023-02-15 01:14:32,907] torch._dynamo.convert_frame: [INFO] GUARDS:                                         
 -                                                                                                            
            local 'm' NN_MODULE                                                                               
            {                                                                                                 
                'guard_types': ['ID_MATCH'],                                                                                  'code': ['___check_obj_id(m, 140215066080400)'],                                              
                'obj_weakref': <weakref at 0x7f8656059310; to 'TestModule' at 0x7f865d34b490>                 
                'guarded_class': <weakref at 0x7f8669d89f90; to 'type' at 0x76868e0 (TestModule)>             
            }                                                                                                 
                                                                                                              
 -                                                                                                            
            local 'tensor' TENSOR_MATCH                                                                       
            {                            
                'guard_types': ['TENSOR_MATCH'],                                                              
                'code': None,                                                                                                 'obj_weakref': <weakref at 0x7f86560593b0; to 'Tensor' at 0x7f865d3b9770>                     
                'guarded_class': <weakref at 0x7f86687a43b0; to 'torch._C._TensorMeta' at 0x5717b50 (Tensor)> 
            }                                                                                                 
                                                                                                              
 -                                                                                                            
            global '__import_torch_dot_nn_dot_modules_dot_module.torch' FUNCTION_MATCH                        
            {                                                                                                
                'guard_types': None,                                                                          
                'code': None,                                                                                
                'obj_weakref': None                                                                                           'guarded_class': None                                                                         
            }                                                                                                 
                                                                                                              
 -                                                                                                            
            global '__import_torch_dot_nn_dot_modules_dot_module._global_forward_hooks' DICT_KEYS             
            {                                                                                                 
                'guard_types': ['DICT_KEYS'],                                                                 
                'code': ['___check_type_id(__import_torch_dot_nn_dot_modules_dot_module._global_forward_hooks,
 7305376)', 'set(__import_torch_dot_nn_dot_modules_dot_module._global_forward_hooks.keys()) == set()'],       
                'obj_weakref': None                                                                           
                'guarded_class': <weakref at 0x7f8771016db0; to 'type' at 0x6f78a0 (OrderedDict)>            
            }                                                                                                                                                                                                                -                                                                                                           
            global '__import_torch_dot_nn_dot_modules_dot_module._global_backward_hooks' DICT_KEYS            
            {                                                                                                 
                'guard_types': ['DICT_KEYS'],                                                                 
                'code': ['___check_type_id(__import_torch_dot_nn_dot_modules_dot_module._global_backward_hooks
, 7305376)', 'set(__import_torch_dot_nn_dot_modules_dot_module._global_backward_hooks.keys()) == set()'],     
                'obj_weakref': None                                                                           
                'guarded_class': <weakref at 0x7f8771016db0; to 'type' at 0x6f78a0 (OrderedDict)>             
            }                                                                                                 
                                                                                                              
 -                                                                                                           
            global '__import_torch_dot_nn_dot_modules_dot_module._global_forward_pre_hooks' DICT_KEYS         
            {                                                                                                 
                'guard_types': ['DICT_KEYS'],                                                                 
                'code': ['___check_type_id(__import_torch_dot_nn_dot_modules_dot_module._global_forward_pre_ho
oks, 7305376)', 'set(__import_torch_dot_nn_dot_modules_dot_module._global_forward_pre_hooks.keys()) == set()']
,                                                                                                             
                'obj_weakref': None                                                                           
                'guarded_class': <weakref at 0x7f8771016db0; to 'type' at 0x6f78a0 (OrderedDict)>             
            }                                                                                                 
                                                                                                              
 -                                                                                                            
            global '__import_torch_dot_nn_dot_modules_dot_module._global_backward_pre_hooks' DICT_KEYS        
            {                                                                                                 
                'guard_types': ['DICT_KEYS'],                                                                                 'code': ['___check_type_id(__import_torch_dot_nn_dot_modules_dot_module._global_backward_pre_h
ooks, 7305376)', 'set(__import_torch_dot_nn_dot_modules_dot_module._global_backward_pre_hooks.keys()) == set()
'],                                                                                                           
                'obj_weakref': None                                                                           
                'guarded_class': <weakref at 0x7f8771016db0; to 'type' at 0x6f78a0 (OrderedDict)>             
            }                                                                                                 
                                                                                                              
 -                                                                                                            
            local_nn_module 'm._forward_hooks' DICT_KEYS                                                   
            {                                                                                                                 'guard_types': None,                                                                          
                'code': None,                                                                                 
                'obj_weakref': None                                                                           
                'guarded_class': None                                                                         
            }                                                                                                 
                                                                                                              
 -                                                                                                           
            local_nn_module 'm._backward_hooks' DICT_KEYS                                                     
            {                                                                                                
                'guard_types': None,                                                                                          'code': None,                                                                                 
                'obj_weakref': None                                                                           
                'guarded_class': None                                                                         
            }                                                                                                 
                                                                                                              
 -                                                                                                            
            local_nn_module 'm._forward_hooks[0]' FUNCTION_MATCH                                              
            {                                                                                                 
                'guard_types': None,                                                                          
                'code': None,                                                                                 
                'obj_weakref': None                                                                          
                'guarded_class': None                                                                                     }                                                                                                                                                                                                              
 -                                                                                                            
            local_nn_module 'm._forward_pre_hooks' DICT_KEYS                                                  
            {                                                                                                 
                'guard_types': None,                                                                          
                'code': None,                                                                                 
                'obj_weakref': None                                                                           
                'guarded_class': None                                                                         
            }                                                                                                 
                                                                                                              
 -                                                                                                           
            local_nn_module 'm._backward_pre_hooks' DICT_KEYS                                                 
            {                                                                                                 
                'guard_types': None,                                                                          
                'code': None,                                                                                 
                'obj_weakref': None                                                                           
                'guarded_class': None                                                                         
            }                                                                                                 
                                                                                                              
 -                                                                                                            
            local_nn_module 'm._forward_hooks_with_kwargs' DICT_KEYS                                          
            {                                                                                                 
                'guard_types': None,                                                                          
                'code': None,                                                                                 
                'obj_weakref': None                                                                                           'guarded_class': None                                                                         
            }                                                                                                 
                                
```

Reply:
yeah `"hooks"` is killing you on DICT_KEYS, which also seems entirely spurious - since you don't care about new hooks added, right? You just want to protect that the functions that did get called are safe. 

I would recommend either adding a proper source for this, or doing something along the lines of  `and ("hooks" not in guard.name if FUNCTION_MATCH else True)`

Reply:
Ah, hmm, on second thought, maybe you do care, if the list of hooks changed? Perhaps we need to remove the dict_key hooks, but add a constant match on number of hooks?

Reply:
ok i guess i could buy that the actual 'key' values do not matter.  We should care about 
 - the actual function matches
 - the order of the functions (in the hook dict, which is an Odict)

Probably OK if someone deleted a hook and added back the same hook with a new key? although ... why